cpp/dspace/doi.cc

```diff
@@ -0,0 +1,98 @@
+    if (EmailSender::SimplerSendEmail("no-reply@ub.uni-tuebingen.de", notification_email_addresses, email_subject, email_message,
+                                      EmailSender::VERY_HIGH)
+        > 299)
+        LOG_ERROR("Failed to send the DOI notification email!");
+}
+
+void UpdateItem(DbConnection * const db_writer, const std::string doi_link, const std::string publication_id) {
+    db_writer->queryOrDie("UPDATE tuefind_publications SET doi_link = " + db_writer->escapeAndQuoteString(doi_link)
+                          + ",doi_notification = CURDATE() WHERE id= " + publication_id + "");
+}
+
+bool DownloadAndUpdate(DbConnection * const db_writer, const std::string external_document_id, const std::string publicationID) {
+    const std::string DOWNLOAD_URL(DspaceServerURL + external_document_id + "/metadata");
+
+    Downloader downloader(DOWNLOAD_URL, Downloader::Params(), 15 * 1000);
+    if (downloader.anErrorOccurred()) {
+        LOG_WARNING("Error while downloading data for id : " + downloader.getLastErrorMessage());
+        return false;
+    }
+
+    const HttpHeader http_header(downloader.getMessageHeader());
+    if (http_header.getStatusCode() != 200) {
+        LOG_WARNING("DOI returned HTTP status code " + std::to_string(http_header.getStatusCode()) + "! for item id: " + publicationID);
+        return false;
+    }
+
+    const std::string &json_document(downloader.getMessageBody());
+    std::shared_ptr<JSON::JSONNode> full_tree;
+    JSON::Parser parser(json_document);
+    if (not parser.parse(&full_tree))
+        LOG_ERROR("failed to parse JSON (" + parser.getErrorMessage());
+
+    const std::shared_ptr<const JSON::ArrayNode> top_node_array(JSON::JSONNode::CastToArrayNodeOrDie("full_tree", full_tree));
+
+    std::string PublicationTitle;
+
+    for (auto item_iter : *top_node_array) {
+        const std::string title_key(JSON::LookupString("/key", item_iter, ""));
+        const std::string title_value(JSON::LookupString("/value", item_iter, ""));
+        if (PublicationTitleKey == title_key) {
+            PublicationTitle = title_value;
+        }
+    }
+
+    for (auto item_iter : *top_node_array) {
+        const std::string one_key(JSON::LookupString("/key", item_iter, ""));
+        const std::string DOILink(JSON::LookupString("/value", item_iter, ""));
+        if (DOIKey == one_key) {
+            std::cout << "Processing: " << DOILink << '\n';
+            UpdateItem(db_writer, DOILink, publicationID);
+            SendNotificationsForDOI("DOI link notification",
+                                    "DOI link: " + DOILink + " successfully generated for publication " + PublicationTitle);
+        }
+    }
+
+    return true;
+}
+
+int main() {
+    auto db_reader((DbConnection::VuFindMySQLFactory()));
+
+    if (not VuFind::GetTueFindFlavour().empty()) {
+        auto db_writer(DbConnection::VuFindMySQLFactory());
```

Comment:
I don't think we need 2 different DB connections to the same database (VuFind)

Reply:
(or maybe we do since we iterate & write at the same time... if that's the case, its ok)